src/module.c

```diff
@@ -1074,213 +1072,508 @@ int RM_CreateSubcommand(RedisModuleCommand *parent, const char *name, RedisModul
+                cmd->key_specs[j].find_keys_type = KSPEC_FK_UNKNOWN;
+                break;
+            case REDISMODULE_KSPEC_FK_RANGE:
+                cmd->key_specs[j].find_keys_type = KSPEC_FK_RANGE;
+                cmd->key_specs[j].fk.range.lastkey = info->key_specs[j].fk.range.lastkey;
+                cmd->key_specs[j].fk.range.keystep = info->key_specs[j].fk.range.keystep;
+                cmd->key_specs[j].fk.range.limit = info->key_specs[j].fk.range.limit;
+                break;
+            case REDISMODULE_KSPEC_FK_KEYNUM:
+                cmd->key_specs[j].find_keys_type = KSPEC_FK_KEYNUM;
+                cmd->key_specs[j].fk.keynum.keynumidx = info->key_specs[j].fk.keynum.keynumidx;
+                cmd->key_specs[j].fk.keynum.firstkey = info->key_specs[j].fk.keynum.firstkey;
+                cmd->key_specs[j].fk.keynum.keystep = info->key_specs[j].fk.keynum.keystep;
+                break;
+            default:
+                serverAssert(0);
+            }
+        }
+        populateCommandLegacyRangeSpec(cmd);
+        populateCommandMovableKeys(cmd);
+    }
+
+    if (info->args) {
+        cmd->args = moduleCopyCommandArgs(info->args);
+        /* Populate arg.num_args with the number of subargs, recursively */
+        cmd->num_args = populateArgsStructure(cmd->args);
+    }
     return REDISMODULE_OK;
 }
 
-int moduleSetCommandKeySpecBeginSearch(RedisModuleCommand *command, int index, keySpec *spec) {
-    struct redisCommand *cmd = command->rediscmd;
+/* Returns 1 if the command info is valid and 0 otherwise. */
+static int moduleValidateCommandInfo(const RedisModuleCommandInfo *info) {
 
-    if (index >= cmd->key_specs_num)
-        return REDISMODULE_ERR;
+    /* No validation for the fields summary, complexity and since (strings or
+     * NULL), hints (NULL-terminated array of strings), arity (any integer). */
 
-    cmd->key_specs[index].begin_search_type = spec->begin_search_type;
-    cmd->key_specs[index].bs = spec->bs;
+    /* History: If since is set, changes must also be set. */
+    if (info->history) {
+        for (size_t j = 0; info->history[j].since; j++) {
+            if (!info->history[j].changes) return 0;
+        }
+    }
 
-    return REDISMODULE_OK;
-}
+    /* Key specs. */
+    if (info->key_specs) {
+        for (size_t j = 0; info->key_specs[j].begin_search_type; j++) {
+            if (j >= INT_MAX) return 0; /* redisCommand.key_specs_num is an int. */
 
-int moduleSetCommandKeySpecFindKeys(RedisModuleCommand *command, int index, keySpec *spec) {
-    struct redisCommand *cmd = command->rediscmd;
+            if (info->key_specs[j].flags & ~(REDISMODULE_CMD_KEY_WRITE |
+                                             REDISMODULE_CMD_KEY_READ |
+                                             REDISMODULE_CMD_KEY_SHARD_CHANNEL |
+                                             REDISMODULE_CMD_KEY_INCOMPLETE))
+                return 0; /* Some undefined flag is set. */
```

Comment:
An alternative to listing all the flags here is to `#define _REDISMODULE_CMD_NEXT (1<<5)` and then use `_REDISMODULE_CMD_NEXT - 1` as the set of all flags. What's the reviewer's preference?